design/crd-configstatus.md

```diff
@@ -0,0 +1,348 @@
+
+```bash
+kubectl get configurationstatus config-status -n metallb-system
+# NAME            READY   REASON
+# config-status   True    AllComponentsReady
+
+# NAME            READY   REASON
+# config-status   False   ComponentFailing (summary which components failed and how)
+```
+
+Secondary interaction
+
+```bash
+# Wait for MetalLB to be ready (common in automation/CI)
+kubectl wait --for=condition=Ready configurationstatus/config-status -n metallb-system --timeout=60s
+
+# Wait for Ready=False (useful in negative tests)
+kubectl wait --for=condition=Ready=False configurationstatus/config-status -n metallb-system --timeout=10s
+
+kubectl get configurationstatus config-status -n metallb-system -o yaml | grep -A 10 "controller/poolReconcilerValid"
+# Shows the actual error message in the condition 
+
+# 3. Fix the configuration and wait for recovery
+kubectl apply -f fixed-config.yaml
+kubectl wait --for=condition=Ready configurationstatus/config-status -n metallb-system --timeout=30s
+```
+
+### Condition Structure per Ready (User facing)
+
+This is calculated centrally after any other condition changes:
+
+- **Type**:  Ready ( to make kubectl wait works)
+- **Status**: `True` (valid) or `False` (invalid)
+- **Reason**:  join list of which condition failed
+- **Message**: Error message  summary
+
+### Condition Structure per Controller (Visible to user, but internal implementation)
+
+Each controller reports its status as a separate condition:
+
+- **Type**: Controller identifier with "Valid" suffix (e.g., "controller/poolReconcilerValid", "speaker-node1/configReconcilerValid")
+- **Status**: `True` (valid) or `False` (invalid)
+- **Reason**: Determined by the following hierarchy:
+  1. If configuration error occurred: `ConfigError`
+  2. Otherwise: The SyncState value
+     - `SyncStateSuccess`: Configuration is valid and applied (Status=True)
+     - `SyncStateReprocessAll`: Configuration applied, requires service reload (Status=True)
+     - `SyncStateError`: Sync operation failed, transient error, will retry (Status=False)
+     - `SyncStateErrorNoRetry`: Sync operation failed, non-transient error, no retry (Status=False)
+- **Message**: Error message during reconcile, if any (empty if no error)
+
+### Which Controller(Reconcilers) Reports
+
+Any Reconciler can report each condition every time reconciler loop runs as long as it
+knows the global config and has patch RBAC. This can be achieved through a
+defer that should never fail. There is no watch on any ConfigStatus resource,
+**only patches its own condition**. 
+
+Even we can add all reconcilers (which should be ok), we must have
+
+* Controller (Deployment - single instance)
+    * PoolReconciler:  Condition: `controller/poolReconcilerValid`
+    * ServiceReconciler:  Condition: `controller/serviceReconcilerValid`
+
+* Speaker (DaemonSet - one pod per node)
+    * ConfigReconciler: Condition: `speaker-<nodename>/configReconcilerValid`
```

Comment:
So this is the reconciler that we only care?

```
	cfg, err := toConfig(resources, r.ValidateConfig)
	if err != nil {
		configStale.Set(1)
		syncResult = SyncStateError
		level.Error(r.Logger).Log("controller", "ConfigReconciler", "error", "failed to parse the configuration", "error", err)
		syncError = fmt.Errorf("failed to parse configuration: %w", err)
		return ctrl.Result{}, nil
	}

````

https://github.com/metallb/metallb/blob/621e1bceeff96f07b6520cdc780fd28d90533a6f/internal/k8s/controllers/config_controller.go#L137